src/taskorchestrationexecutor.ts

```diff
@@ -0,0 +1,493 @@
+                this.output = generatorResult.value;
+                return;
+            } else if (generatorResult.value instanceof Task) {
+                // The generator yielded another task.
+                newTask = generatorResult.value.internalTask;
+            }
+        } catch (exception) {
+            // The generator threw an exception
+            this.exception = exception;
+        }
+
+        if (newTask !== undefined) {
+            // The generator returned an already-completed task,
+            // so we try to run the user code again.
+            this.currentTask = newTask;
+            if (newTask.state !== TaskState.Running) {
+                this.tryResumingUserCode();
+            } else {
+                // The task hasn't completed, we add it to the open (incomplete) task list
+                this.trackOpenTask(newTask);
+                // We only keep track of actions from user-declared tasks, not from
+                // tasks generated internally to facilitate history-processing.
+                if (this.currentTask instanceof ProperTask) {
+                    this.addToActions(this.currentTask.actionObj);
+                }
+            }
+        }
+    }
+
+    /**
+     * @hidden
+     * Add an action to the user-defined actions list.
+     * It ignores the request if the orchestrator has already
+     * signaled a "ContinueAsNew" operation.
+     *
+     * @param action
+     *  User-defined action to track
+     */
+    public addToActions(action: IAction): void {
+        if (!this.willContinueAsNew) {
+            this.actions.push(action);
+        }
+    }
+
+    /**
+     * @hidden
+     * Tracks this task as waiting for completion.
+     * In the process, it assigns the task an ID if it doesn't have one already.
+     *
+     * @param task
+     *  Task to add to open tasks or open events list
+     */
+    public trackOpenTask(task: InternalOnlyTask | ProperTask): void {
+        // TODO: should there be a check for the task status as RUNNING?
+
+        // The open tasks and open events objects only track singular tasks, not compound ones.
+        // Therefore, for a compound task, we recurse down to its inner sub-tasks add
+        // record all singular tasks.
+        if (task instanceof CompoundTask) {
+            for (const child of task.children) {
+                this.trackOpenTask(child);
+            }
+        } else {
+            if (task.id === false) {
+                // The task needs to be given an ID and then stored.
+                task.id = this.sequenceNumber++;
+                this.openTasks[task.id] = task;
+            } else if (task.actionObj instanceof WaitForExternalEventAction) {
+                // The ID of `WaitForExternalEvent` tasks is the name of
+                // the external event. Given that multiple `WaitForExternalEvent`
+                // tasks can be led for the same event name at once, we need
```

Comment:
"led"?

Reply:
Yeah, it's not the right verb. I'll try with "can refer to"  or something like that.